Document polls (#218)
discord-userdoccers/discord-userdoccers: pages/resources/poll.mdx

| layout_type       | integer                                                    | The [layout type](#layout-type) of the poll         |
| results?          | [poll results](#poll-results-object-structure) object      | The results of the poll                             |

`expiry` is marked as nullable to support non-expiring polls in the future, but all polls have an expiry currently.

###### Poll Object Example

```json
{
  "question": {
    "text": "Dolfies"
  },
  "answers": [
    {
      "answer_id": 1,
      "poll_media": {
        "text": "Alien 1"
      }
    },
    {
      "answer_id": 2,
      "poll_media": {
        "text": "Alien 2",
        "emoji": {
          "id": null,
          "name": "👽"
        }
      }
    },
    {
      "answer_id": 3,
      "poll_media": {
        "text": "Alien 3",
        "emoji": {
          "id": "1235169043636424757",
          "name": "dolfies"
        }
      }
    }
  ],
  "expiry": "2024-05-02T10:00:02.039342+00:00",
  "allow_multiselect": false,
  "layout_type": 1,
  "results": {
    "answer_counts": [],

>     "is_finalized": false

Reviewer: I think an example with results would be better

| answers           | array[[poll answer](#poll-answer-object-structure) object] | Each of the answers available in the poll, up to 10                  |
| duration          | integer                                                    | Number of hours the poll should be open for, up to 7 days            |
| allow_multiselect | boolean                                                    | Whether a user can select multiple answers                           |
| layout_type?      | integer                                                    | The [layout type](#layout-type) of the poll. Defaults to... DEFAULT! |

###### Example Poll Create Request Object

```json
{
  "question": {
    "text": "Dolfies"
  },
  "answers": [
    {
      "poll_media": {
        "text": "Alien 1"
      }
    },
    {
      "poll_media": {
        "text": "Alien 2",
        "emoji": {
          "name": "👽"
        }
      }
    },
    {
      "poll_media": {
        "text": "Alien 3",
        "emoji": {
          "id": "1235169043636424757",
          "name": ""
        }
      }
    }
  ],
  "allow_multiselect": false,
  "duration": 24,
  "layout_type": 1
}
```

> ### Layout Type

Reviewer: ```suggestion
###### Layout Type
```

We might have different layouts for polls in the future.

> For now though, this number will be 1.

Reviewer: ```suggestion
Different layouts for polls will come in the future. For now though, this value will always be `DEFAULT`.
```
We aren't discord :)

| Type                   | ID    | Description                            |
| ---------------------- | ----- | -------------------------------------- |
| DEFAULT                | 1     | The, uhm, default layout type.         |

> | ~~IMAGE_ONLY_ANSWERS~~ | ~~2~~ | ~~Poll answers can have only images.~~ |

Reviewer: ```suggestion
| Value    | Name                   |  Description                            |
| ----- | ---------------------- | -------------------------------------- |
| 1 | DEFAULT                | The default layout type         |
| ~~2~~ | ~~IMAGE_ONLY_ANSWERS~~ | ~~Poll answers can have only images~~ |
```

### Poll Media Object

The poll media object is a common object that backs both the question and answers.

> The intention is that it allows us to extensibly add new ways to display things in the future.

Reviewer: ```suggestion
```

For now, `question` only supports `text`, while answers can have an optional `emoji`.

###### Poll Media Object Structure

| Field  | Type                                           | Description            |
| ------ | ---------------------------------------------- | ---------------------- |

> | text?  | string                                         | The text of the field  |

Reviewer: ```suggestion
| text?  | string                                         | The text of the field (max 300 characters for question, 55 characters for answer)  |
```

| emoji? | partial [emoji](/resources/emoji#emoji-object) | The emoji of the field |

> `text` should always be non-null for both questions and answers, but please do not depend on that in the future.

Reviewer: Should be a footnote

> The maximum length of `text` is 300 for the question, and 55 for any answer.

Reviewer: ```suggestion
```

> When creating a poll answer with an emoji, one only needs to send either the `id` (custom emoji) or `name` (default emoji) as the only field.

Reviewer: Should be a footnote

### Poll Answer Object

The `answer_id` is a number that labels each answer.
As an implementation detail, it currently starts at 1 for the first answer and goes up sequentially.
We recommend against depending on this sequence.

Currently, there is a maximum of 10 answers per poll.

###### Poll Answer Object Structure

| Field         | Type                                              | Description            |
| ------------- | ------------------------------------------------- | ---------------------- |
| answer_id ^1^ | integer                                           | The ID of the answer   |
| poll_media    | [poll media](#poll-media-object-structure) object | The data of the answer |

> ^1^ Only sent as part of responses from Discord's API/Gateway.

Reviewer: ```suggestion
^1^ When sending, this field is optional.
```

### Poll Results Object

In a nutshell, this contains the number of votes for each answer.

The `results` field may be not present in certain responses where, as an implementation detail, we do not fetch the poll results in our backend.
This should be treated as "unknown results", as opposed to "no results". You can keep using the results if you have previously received them through other means.

Also due to the intricacies of counting at scale, while a poll is in progress the results may not be perfectly accurate.

> They usually are accurate, and shouldn't deviate significantly -- it's just difficult to make guarantees.

Reviewer: ```suggestion
They usually are accurate, and shouldn't deviate significantly—it's just difficult to make guarantees.
```

To compensate for this, after a poll is finished there is a background job which performs a final, accurate tally of votes.
This tally concludes once `is_finalized` is `true`. Polls that have ended will also always contain results.

If `answer_counts` does not contain an entry for a particular answer, then there are no votes for that answer.

> ###### Poll Results Object Structure

Reviewer: ```suggestion
###### Poll Results Structure
```

| Field         | Type                                                                   | Description                                   |
| ------------- | ---------------------------------------------------------------------- | --------------------------------------------- |
| is_finalized  | boolean                                                                | Whether the votes have been precisely counted |
| answer_counts | array[[poll answer count](#poll-answer-count-object-structure) object] | The counts for each answer                    |

> ###### Poll Answer Count Object Structure

Reviewer: ```suggestion
###### Poll Answer Count Structure
```

| Field    | Type    | Description                                    |
| -------- | ------- | ---------------------------------------------- |

> | id       | integer | The `answer_id`                                |

Reviewer: ```suggestion
| id       | integer | The ID of the answer                               |
```

| count    | integer | The number of votes for this answer            |
| me_voted | boolean | Whether the current user voted for this answer |

## Endpoints

For creating a poll, see [Create Message](/resources/message#create-message). After creation, the poll message cannot be edited.

Apps are not allowed to vote on polls. No rights! :)

> 

Reviewer: ```suggestion
```

<RouteHeader method="GET" url="/channels/{channel.id}/polls/{message.id}/answers/{answer.id}">
  Get Answer Voters
</RouteHeader>

Get a list of users that voted for this specific answer.

###### Query String Params

| Field  | Type      | Description                           | Default |
| ------ | --------- | ------------------------------------- | ------- |
| after? | snowflake | Get users after this user ID          | absent  |

> | limit? | integer   | Max number of users to return (1-100) | 25      |

Reviewer: ```suggestion
| Field  | Type      | Description                           |
| ------ | --------- | ------------------------------------- |
| after? | snowflake | Get users after this user ID          |
| limit? | integer   | Max number of users to return (1-100, default 25)
```
